### PR TITLE
[Fix] Try-state feature-gated for BagsList

### PR DESCRIPTION
**Pull request by ruseinov**
fixes https://github.com/paritytech/substrate/issues/13293

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
Should be fine as long as it compile and manages to guard `SortedListProvider::try_state` behind the feature ✅

**Comment by ruseinov** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352323

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352370

**Comment by ruseinov** [comment]
@kianenigma this seems to look good, however I had to feature-gate `remote-tests` `try_state` call for bagsList due to the fact that `try_state` is `try-runtime` only now. It did not seem right to have `do_try_state` as a public method.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352463

**Comment by Gesge** [comment]
What's up guy?This is my automatic reply~ With the purpose of enjoying the holiday,I did this auto-reply~Please wait a while but no too long.As soon as I saw it,turn back to your letter~~~

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352499

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352500

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352506

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352501

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352554

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352556

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352555

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352651

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352649

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352650

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352743

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352745

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352744

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352794

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352793

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352792

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352842

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352843

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352841

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352891

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352892

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2352890

File: frame/bags-list/remote-tests/Cargo.toml [hunk 1]
```diff
@@ -15,7 +15,7 @@ targets = ["x86_64-unknown-linux-gnu"]
 [dependencies]
 # frame
 pallet-staking = { path = "../../staking", version = "4.0.0-dev" }
-pallet-bags-list = { path = "../../bags-list", version = "4.0.0-dev" }
+pallet-bags-list = { path = "../../bags-list", version = "4.0.0-dev", features = ["fuzz"] }
```
Comment (ruseinov) [inline frame/bags-list/remote-tests/Cargo.toml:18]: @kianenigma this is the change I am not a fan of, perhaps we can rename this feature to something like remote-tests or external-tests. Or keep as is.

File: frame/bags-list/src/list/mod.rs [hunk 2]
```diff
@@ -790,6 +790,7 @@ impl<T: Config<I>, I: 'static> Bag<T, I> {
 	}
 
 	/// Check if the bag contains a node with `id`.
+	#[allow(dead_code)]
```
Comment (kianenigma) [inline frame/bags-list/src/list/mod.rs:793]: should be removed?

Comment (kianenigma) [reply frame/bags-list/src/list/mod.rs:793]: or else gated with the right features.

File: frame/bags-list/remote-tests/Cargo.toml (hunk 1 above)
Comment (kianenigma) [reply frame/bags-list/remote-tests/Cargo.toml:18]: I think it is fine.

File: frame/bags-list/src/list/mod.rs [hunk 3]
```diff
@@ -514,7 +511,9 @@ impl<T: Config<I>, I: 'static> List<T, I> {
 	/// * length of this list is in sync with `ListNodes::count()`,
 	/// * and sanity-checks all bags and nodes. This will cascade down all the checks and makes sure
 	/// all bags and nodes are checked per *any* update to `List`.
-	pub(crate) fn try_state() -> Result<(), &'static str> {
+	#[cfg(any(test, feature = "try-runtime", feature = "fuzz"))]
+	#[allow(dead_code)]
```
Comment (kianenigma) [inline frame/bags-list/src/list/mod.rs:515]: When does this end up being unused?

**Comment by ruseinov** [comment]
bot rebase

File: frame/bags-list/src/list/mod.rs (hunk 3 above)
Comment (ruseinov) [reply frame/bags-list/src/list/mod.rs]: fixed that, debugging leftover.

File: frame/bags-list/src/list/mod.rs (hunk 2 above)
Comment (ruseinov) [reply frame/bags-list/src/list/mod.rs]: right, done!

**Comment by ruseinov** [comment]
bot rebase

File: frame/bags-list/src/mock.rs [hunk 4]
```diff
@@ -148,7 +148,7 @@ impl ExtBuilder {
 	pub fn build_and_execute(self, test: impl FnOnce() -> ()) {
 		self.build().execute_with(|| {
 			test();
-			List::<Runtime>::try_state().expect("Try-state post condition failed")
+			List::<Runtime>::do_try_state().expect("Try-state post condition failed")
```
Comment (melekes) [inline frame/bags-list/src/mock.rs:151]:
```suggestion
			List::<Runtime>::do_try_state().expect("do_try_state post condition failed")
```

**Review by melekes — COMMENTED** [review]
lgtm

File: frame/bags-list/src/mock.rs (hunk 4 above)
Comment (ruseinov) [reply frame/bags-list/src/mock.rs:151]: Ideologically it is still `try-state`, `do_try_state` is just a helper. But why not

**Review by gilescope — APPROVED** [review]

**Comment by ruseinov** [comment]
bot merge

**Pull request merged**